Hannah-Sten/TeXiFy-IDEA
# Listing label support

Fixes #1217 (including verbatim)

#### Summary of additions and changes

* fixed broken LatexFigureNotRferenced inspection
* implement missing label inspection for lstlistings and other parameter labeled environments
* implement missing label quickfix for lstlistings and other parameter labeled environments
* implmeent goto reference and autocompletion for environment labels
* change hasLabel to getLabel for environments (also returns the label name if present)
* change LatexPsiFactory to LatexPsiHelper. The new class also helps with certain modifications. Unfortunately the changes in this class are not backwards compatible, but as I found myself adding more and more PSI modifications in the quickfixes I decided to change the factory to support clients also with modifying the tree.

#### How to test this pull request

* see the included tests
* try to add environments (figures, listings, verbatims, etc) in a testfile and try the suggested quickfixes 

#### General notes

* I decided to add the tests inline this time because I think it makes the tests more obvious. Which version do you like better?
* As already mentioned in Gitter, the LabelConvention quickfix basically could perform a rename refactoring if one was present yet. Once we have refactoring support for labels, the quickfix can be simplified.

#### Wiki


- [x] Updated the wiki: https://github.com/Hannah-Sten/TeXiFy-IDEA/wiki/Conventions#missing-labels
